modules/axtask/src/wait_queue.rs

```diff
@@ -81,23 +106,39 @@ impl WaitQueue {
     where
         F: Fn() -> bool,
     {
+        let mut rq = current_run_queue::<NoPreemptIrqSave>();
         loop {
-            let mut rq = RUN_QUEUE.lock();
+            let mut wq = self.queue.lock();
             if condition() {
                 break;
             }
-            rq.block_current(|task| {
-                task.set_in_wait_queue(true);
-                self.queue.lock().push_back(task);
-            });
+            let curr = crate::current();
+            assert!(curr.is_running());
+            assert!(!curr.is_idle());
+
+            debug!("{} push to wait queue on wait_until", curr.id_name());
+
+            // we must not block current task with preemption disabled.
+            // Current expected preempt count is 2.
+            // 1 for `NoPreemptIrqSave`, 1 for wait queue's `SpinNoIrq`.
+            #[cfg(feature = "preempt")]
+            assert!(curr.can_preempt(2));
+            wq.push_back(curr.clone());
+
+            curr.set_state(TaskState::Blocked);
```

Comment:
Try to put all task state manipulation to `run_queue.rs`

Reply:
Tring, we have to make sure task's state is set as `Blocked` with holding the lock of wait queue!